Add `Manage this user` option in user-info popover for admins.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This commit adds "Mange this user" option in the user-info popover which opens
the user-info modal.
    
We show a spinner on submit button in this case as modal is not closed immediately
and thus we need some indicator to show that the task is in progress. There is no spinner
on submit button in the modal opened from "Users" section of organization settings.
    
Error handling for this case is different than when the modal is opened from "Users"
section of organization settings because there is no overlay in the background of
modal in this case.
    
In this case, we show error inside the modal and do not close it and in case the change
is completed successfully we just close the modal without showing any message.
    
Fixes part of #18944.
 <!-- How have you tested? -->

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
Modal when the change is failed and server returns error.
![change-user-info-failed](https://user-images.githubusercontent.com/35494118/143597953-a3b47339-a752-4df6-ab42-9704cce77b8d.gif)

User information updated successfully
![change-user-info](https://user-images.githubusercontent.com/35494118/143597963-bdb6c144-8505-4451-ab8e-0d0eae0c552e.gif)

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented:
Thanks for working on this!  A few quick thoughts:
* It looks like the new edit icon is a bit off-center vertically; maybe it needs to be a pixel or two down?  (I think what's going on is this font-awesome icon is just not centered without its box.)
* I feel like it's a bit weird having so many horizontal lines; maybe we should not add a newline between this and the "Mute user"?  I think that's the cleanest visual option barring some larger reorganization.
* While we're at this, we should probably change the title of the dialog to "Manage user" rather than "Change user info and role" (obviously applying in both places this appears).
* Also, I notice we're not using `simplebar` for the scrollbars in this widget; I expect that's an accidental side effect of the Micromodal migration (we just don't have many Modals that might be tall enough to generate a scrollbar).
* I guess what remains for completing the original issue is adding a "Deactivate user" button to the modal?

sahil839 commented:
- Fixed the icon alignment. Attaching the screenshot -
![Screenshot from 2021-11-28 19-56-00](https://user-images.githubusercontent.com/35494118/143908151-c65bf48a-f18b-491a-89c8-f7c8daafee6e.png)

- Removed the horizontal lines and updated the heading of modal.
- Added commit to use simplebar.

sahil839 commented:
Also added a commit for `Deactivate this user` option.

![deactivate-user](https://user-images.githubusercontent.com/35494118/143908767-b67d3df2-452c-40cd-b7f4-6b7de74bc6ea.gif)
![deactivate_failed](https://user-images.githubusercontent.com/35494118/143908779-c9e3b6bf-53bb-4976-85f6-8cbe2575f8d8.gif)

alya commented:
I played around with this PR for a few minutes, and it looks good to me!

timabbott commented:
I thought our plan was to add "Deactivate this user" to the "Manage this user" widget, not the normal user profile popover?  It feels kinda in your face for admins to see that whenever they look at another user.

alya commented:
Ah, yeah, I forgot that discussion earlier, but that sounds right!

timabbott commented:
OK, I merged the first 4 commits as the series ending with 4eeee7296e18eb09c52ae29ec4046bbcd13e3b89.

timabbott commented on static/js/settings_users.js:
```diff
@@ -573,15 +573,31 @@ export function show_edit_user_info_modal(user_id, status_field) {
+        if (!from_user_info_popover) {
+            settings_ui.do_settings_change(channel.patch, url, data, status_field);
+            dialog_widget.close_modal();
+            return;
```
Why does the other code path close the modal before `do_settings_change` returns, rather than having that be in a `success` handler?  I think that probably represents a bug / legacy settings feedback style in the existing flow that we should clean up.

sahil839 replied:
It is because the feedback for success/failure is shown in the settings overlay besides the `Users` heading.

timabbott replied:
Yeah, I figured -- I think we probably want the feedback to be in the modal -- that's usually better UX since then you have the opportunity to fix issues if we get an error, etc.

Worth discussing a bit in #design but I think that's the pattern we're trying to move all of our modals to.

sahil839 commented:
> I thought our plan was to add "Deactivate this user" to the "Manage this user" widget, not the normal user profile popover? It feels kinda in your face for admins to see that whenever they look at another user.

Ohh, the issue mentioned to add the option in popover, so I did this.

I am not sure what behavior do we want here, since we open a modal for `Manage user` and then if we would have `Deactivate` button inside we would also need another modal for confirmation and thus we would need to close the existing modal since existing `dialog_widget` code assumes only one modal is opened at a time.
If user then deactivates the user then it is fine, but if user cancels the deactivation from modal, then do we want to show the `Manage user` modal again?

timabbott commented:
I think it'd be reasonable for the "Deactivate" button to close the "Manage user" modal and replace it.  I don't think there's any need to re-show "Manage user" if they decide not to deactivate; the screen underneath the modal will be construction always have a way to get back to it.

sahil839 commented:
Tried adding the `Deactivate` button in the `Manage user` modal, but there is a problem. As we click on the deactivate button, we need to close the `Manage user` modal first and then open deactivation modal, but I am not able to figure out correct way to wait for the first modal to be completely closed, and thus have added `setTimeout` for now, but we would need to change it. If we do not wait for the first modal to be closed, there is an error because we do not allow two modals to be open at same time.

timabbott commented:
> If we do not wait for the first modal to be closed, there is an error because we do not allow two modals to be open at same time. Any suggestions?

Good question.  @ganpa3 do you have ideas for how we might handle that?

ganpa3 commented:
@timabbott @sahil839 I think having an event handler on closing a modal should suffice for now? (#20938)

timabbott commented:
Yeah, I think that should work; posted a quick review on the implementation for that.

sahil839 commented:
Not sure whether that would work. We would want to open the "deactivate" modal only if original modal is closed on clicking `deactivate` button, not on normally closing the modal.

timabbott commented:
OK I'm merging a version of #20938, which should unblock this.

sahil839 commented:
Updated the PR.
![deactivate_user](https://user-images.githubusercontent.com/35494118/155335153-b76db6ad-aa90-4372-baea-44495a2114ff.gif)

timabbott commented:
This is great, merged, thanks @sahil839!  I'll post a few follow-up proposals.

@alya FYI -- we likely have some text we should change in the help center after this.

timabbott commented on static/js/settings_users.js:
```diff
@@ -559,6 +561,36 @@ export function show_edit_user_info_modal(user_id, from_user_info_popover, statu
+                        if (!from_user_info_popover) {
+                            const row = get_user_info_row(user_id);
+                            update_view_on_deactivate(row);
+                        }
```
I think the `update_view_on_deactivate` calls could be eliminated from these `success` handlers entirely if we just modify `user_events.js` to call `update_user_data` when changing the `active` flag on a user; this would be a substantial cleanup.

Also, while we're at this, I noticed that a user deactivated via this flow does not disappear from presence immediately; likely another thing that should be happening in the `user_events.js` handler block for a deactivated user.

alya commented:
> @alya FYI -- we likely have some text we should change in the help center after this.

Good callout! I filed https://github.com/zulip/zulip/issues/21318.
